block/retriever.go

```diff
@@ -6,38 +6,45 @@ import (
 	"github.com/dymensionxyz/gerr-cosmos/gerrc"
 
 	"github.com/dymensionxyz/dymint/da"
+	"github.com/dymensionxyz/dymint/settlement"
 	"github.com/dymensionxyz/dymint/types"
 )
 
-func (m *Manager) ApplyBatchFromSL(daMetaData *da.DASubmitMetaData) error {
-	m.logger.Debug("trying to retrieve batch from DA", "daHeight", daMetaData.Height)
-	batchResp := m.fetchBatch(daMetaData)
+func (m *Manager) ApplyBatchFromSL(slBatch *settlement.Batch) error {
+	m.logger.Debug("trying to retrieve batch from DA", "daHeight", slBatch.MetaData.DA.Height)
+	batchResp := m.fetchBatch(slBatch.MetaData.DA)
 	if batchResp.Code != da.StatusSuccess {
 		return batchResp.Error
 	}
 
-	m.logger.Debug("retrieved batches", "n", len(batchResp.Batches), "daHeight", daMetaData.Height)
+	m.logger.Debug("retrieved batches", "n", len(batchResp.Batches), "daHeight", slBatch.MetaData.DA.Height)
 
 	m.retrieverMu.Lock()
 	defer m.retrieverMu.Unlock()
 
 	var lastAppliedHeight float64
+	blockIndex := 0
 	for _, batch := range batchResp.Batches {
 		for i, block := range batch.Blocks {
+			// We dont apply a block if not included in the block descriptor (adds support for rollback)
+			if blockIndex >= len(slBatch.BlockDescriptors) {
```

Comment:
wouldn't it be cleaner to check `block.Header.Height` against the heights height of `BlockDescriptor.Height` so in that case you don't need to add any index etc?

Reply:
i think the index is necessary anyway, if we want to keep compatiblity with  multiple batches in the same state update

Reply:
although not supported right now...

Reply:
another problem I see now that we only increment the `blockIndex` at the end of the loop but we have `continue` in the middle which will make the block index incorrect..

Reply:
fixed